helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/SoftConstraint.java

```diff
@@ -86,11 +69,11 @@ float getAssignmentNormalizedScore(AssignableNode node, AssignableReplica replic
   }
 
   /**
-   * The default scaler function that squashes any score within (min_score, max_score) to (0, 1);
+   * The default normalize function f(x) = tanh(1/(\alpha * x)) that squashes any score >= 0 to (0, 1]
+   * The hyper parameter determines the converging rate of scores to 0 when X gets larger
    * Child class could override the method and customize the method on its own
-   * @return The MinMaxScaler instance by default
    */
   NormalizeFunction getNormalizeFunction() {
-    return (score) -> (score - getMinScore()) / (getMaxScore() - getMinScore());
+    return score -> (score == 0f? 1f: (float) Math.tanh(1 / (score * this.alpha)));
```

Comment:
So we assume all the soft constraints will use this formula? I don't think it will fit the partition movement constraint.
My suggestion is that, keep the scale (max and min), only change the constraints individually which we are calculating the percentage. Those are the applicable ones.

Reply:
The method is in parent class and child class like partition movement can override and have its own implementation without problems. It's reflected in my latest commit.

Reply:
0.0f / 1.0f?

Reply:
Where is the 0.0f in the numerator coming from?

Reply:
One question, if the score could be any number, how it becomes comparable after Math.tanh(1 / (score * this.alpha))?
I know this formula ensures the return number is between 0, 1. But for example, if ScoreA is between 0 - 100, and ScoreB is between 1000 - 500000, even after the tanh(), how you justify their value is comparable?

This is a real concern, according to your previous code if one cluster has 10K partitions, and later more resource added and the count is 100K, part of your evaluation scores might be completely different regarding the actual scale.

What I'm thinking is that try to ensure the sable scale of all soft constraints. So the alpha needs to be configured according to each constraint separately. For instance, partition count constraint, the alpha should be calculated by a function which takes the estimate partition or total partition number as an input.

This algorithm is very sensitive, so we need to clearly justify all the detail designs. Moreover, any logic other than the configuration part should be treated as the final version at this stage.

Reply:
I agree. But the purpose of this PR is not to define the right alpha for each soft constraint but the scores in the same range. In my understanding, the tuning part will resolve the correct coefficients and alphas if necessary. We don't necessarily need to have comparable scores between soft constraints and it'll always be a problem to argue if A constraint is comparable to B constraint because they're implemented differently and independently. The final tuned "coefficients" or "weights" will take the absolute values into considerations and evens out the uncomparable part. The end result is not to compare "absolute" weight between each soft constraint but to generate the "trustworthy" metrics about the learned coefficients like "R2 scores", "MSE", etc. That tells us e.g 80% chance with this set of weights configured, we could achieve an evenness degree of 0.5 something like that. We could discuss offline if you still think the absolute comparable score numbers matter.

Reply:
Synced with @jiajunwang offline, we agreed to use the "usage percentage" instead of the absolute value of the usage as the input to the normalization method.

For each soft constraint, the "usage percentage" is roughly calculated by 
`usage percent = absolute usage value / estimate max usage`